src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java

```diff
@@ -181,23 +218,149 @@ public Set<Component> getComponents()
         return getComponents(indexes);
     }
 
-    // TODO: Currently returns an empty set for in-memory only implementation
-    static Set<Component> getComponents(Collection<Index> indices)
+    static Set<Component> getComponents(Collection<StorageAttachedIndex> indices)
+    {
+        Set<Component> components = Version.LATEST.onDiskFormat()
+                                                  .perSSTableComponents()
+                                                  .stream()
+                                                  .map(c -> new Component(Component.Type.CUSTOM,
+                                                                          Version.LATEST.fileNameFormatter().format(c, null)))
+                                                  .collect(Collectors.toSet());
+        indices.forEach(index -> components.addAll(index.getComponents()));
+        return components;
+    }
+
+    // This differs from getComponents in that it only returns index components that exist on disk.
+    // It avoids errors being logged by the SSTable.readTOC method when we have an empty index.
+    @VisibleForTesting
+    public static Set<Component> getLiveComponents(SSTableReader sstable, Collection<StorageAttachedIndex> indices)
     {
-        return Collections.emptySet();
+        IndexDescriptor indexDescriptor = IndexDescriptor.create(sstable);
+        Set<Component> components = indexDescriptor.getLivePerSSTableComponents();
+        indices.forEach(index -> components.addAll(indexDescriptor.getLivePerIndexComponents(index.getIndexContext())));
+        return components;
     }
 
     @Override
     public void handleNotification(INotification notification, Object sender)
     {
-        if (notification instanceof MemtableRenewedNotification)
+        // unfortunately, we can only check the type of notification via instanceof :(
+        if (notification instanceof SSTableAddedNotification)
         {
-            indexes.forEach(index -> ((StorageAttachedIndex)index).getIndexContext().renewMemtable(((MemtableRenewedNotification) notification).renewed));
+            SSTableAddedNotification notice = (SSTableAddedNotification) notification;
+
+            // Avoid validation for index files just written following Memtable flush.
+            //TODO Do we need to validate streamed index files?
```

Comment:
Is the performance penalty really significant? It's just the header and footer, so seems reasonable to continue to do it, especially given we're getting these files from another node, etc.